server/test/write.go

```diff
@@ -733,6 +752,335 @@ var writeCommandTests = []writeCommandTest{
 			}},
 		},
 	},
+	{
+		_name: "Delete succeeds even if user field contains a type that is not allowed by the current authorization model",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		tuples: []*openfgapb.TupleKey{
+			{
+				Object:   "org:openfga",
+				Relation: "owner",
+				User:     "impossible:1",
+			},
+		},
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type:      "user",
+				Relations: map[string]*openfgapb.Userset{},
+			},
+			{
+				Type: "org",
+				Relations: map[string]*openfgapb.Userset{
+					"owner": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"owner": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type: "user",
+								},
+							},
+						},
+					},
+				},
+			}},
+		request: &openfgapb.WriteRequest{
+			Deletes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("org:openfga", "owner", "impossible:1"),
+			},
+			},
+		},
+	},
+	{
+		_name: "Write fails if user field contains a type that does not exist",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type: "org",
+				Relations: map[string]*openfgapb.Userset{
+					"owner": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"owner": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type: "user",
+								},
+							},
+						},
+					},
+				},
+			}},
+		request: &openfgapb.WriteRequest{
+			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("org:openfga", "owner", "impossible:1"),
+			},
+			},
+		},
+		err: serverErrors.InvalidWriteInput,
+	},
+	{
+		_name: "Write fails if user field contains a type that is not allowed by the authorization model (which only allows group:...)",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type:      "user",
+				Relations: map[string]*openfgapb.Userset{},
+			},
+			{
+				Type: "group",
+				Relations: map[string]*openfgapb.Userset{
+					"member": {Userset: &openfgapb.Userset_This{}},
+				},
+			},
+			{
+				Type: "document",
+				Relations: map[string]*openfgapb.Userset{
+					"reader": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"reader": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type: "group",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		request: &openfgapb.WriteRequest{
+			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("document:budget", "reader", "user:abc"),
```

Comment:
Can we please add a test case that makes sure the write with a user value such as `group:eng#member` is also invalid here. Since `reader` is restricted to `group`, a write such as `document:budget, reader, group:eng#member` is also invalid.